core/src/main/scala/org/apache/spark/scheduler/EventLogFileWriters.scala

```diff
@@ -0,0 +1,444 @@
+ * writer doesn't check the size on file being open for write: the writer tracks the count of bytes
+ * being written currently.
+ *
+ * For metadata files, the class will leverage zero-byte file, as it provides minimized cost.
+ *
+ * The writer leverages the following configurable parameters:
+ *   spark.eventLog.rollLog - Whether rolling over event log files is enabled.
+ *   spark.eventLog.rollLog.maxFileSize - The max size of event log file to be rolled over.
+ */
+class RollingEventLogFilesWriter(
+    appId: String,
+    appAttemptId : Option[String],
+    logBaseDir: URI,
+    sparkConf: SparkConf,
+    hadoopConf: Configuration)
+  extends EventLogFileWriter(appId, appAttemptId, logBaseDir, sparkConf, hadoopConf) {
+
+  import RollingEventLogFilesWriter._
+
+  private val eventFileMaxLengthKiB = sparkConf.get(EVENT_LOG_ROLLED_EVENT_LOG_MAX_FILE_SIZE)
+
+  private val logDirForAppPath = getAppEventLogDirPath(logBaseDir, appId, appAttemptId)
+
+  // Only defined if the file system scheme is not local
+  private var hadoopDataStream: Option[FSDataOutputStream] = None
+  private var countingOutputStream: Option[CountingOutputStream] = None
+  private var writer: Option[PrintWriter] = None
+
+  // seq and event log path will be updated soon in rollNewEventLogFile, which `start` will call
+  private var sequence: Long = 0L
+  private var currentEventLogFilePath: Path = logDirForAppPath
+
+  override def start(): Unit = {
+    requireLogBaseDirAsDirectory()
+
+    if (fileSystem.exists(logDirForAppPath) && shouldOverwrite) {
+      // try to delete the directory
+      fileSystem.delete(logDirForAppPath, true)
+    }
+
+    if (fileSystem.exists(logDirForAppPath)) {
+      // we tried to delete the existing one, but failed
+      throw new IOException(s"Target log directory already exists ($logDirForAppPath)")
+    } else {
+      fileSystem.mkdirs(logDirForAppPath, EventLogFileWriter.LOG_FILE_PERMISSIONS)
+      createAppStatusFile(inProgress = true)
+      rollNewEventLogFile(None)
+    }
+  }
+
+  override def writeEvent(eventJson: String, flushLogger: Boolean = false): Unit = {
+    writer.foreach { w =>
+      val currentLen = countingOutputStream.get.getBytesWritten
+      if (currentLen + eventJson.length > eventFileMaxLengthKiB * 1024) {
+        rollNewEventLogFile(Some(w))
+      }
+    }
+
+    // if the event log file is rolled over, writer will refer next event log file
+
+    // scalastyle:off println
+    writer.foreach(_.println(eventJson))
+    // scalastyle:on println
+    if (flushLogger) {
+      writer.foreach(_.flush())
+      hadoopDataStream.foreach(_.hflush())
+    }
+  }
+
+  private def rollNewEventLogFile(w: Option[PrintWriter]): Unit = {
+    writer.foreach(_.close())
+
+    sequence += 1
+    currentEventLogFilePath = getEventLogFilePath(logDirForAppPath, appId, appAttemptId, sequence,
+      compressionCodecName)
+
+    val streams = initLogFile(currentEventLogFilePath)
```

Comment:
Same comment as before (`val (blah1, blah2) = ...`).

Also, isn't `CountingOutputStream` only needed by this class? So why not add that wrapper here instead of in the parent class?

Reply:
Ah yes that sounds much better! Will apply.